### PR TITLE
ENH: Improve j1 for large inputs

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Towards https://github.com/scipy/scipy/issues/22706

#### What does this implement/fix?
Improve numerical stability for large inputs in $j1$ the same wa as #131 and #140.

#### Additional information
This PR again boils down to rewriting a trigonometrical expression. I used Wolfram Alpha to do that for me [here](https://www.wolframalpha.com/input?i=p+*+cos%28x-%283pi%2F4%29%29+%2B+w+*+q+*+sin%28x-%283pi%2F4%29%29) which yields the following screenshot:

image

Since this is the same pattern as #131 and #140 I went with a cutoff of 10 again. If requested, I can run a benchmark against mpmath to determine a suitable cutoff but I am quite confident that it will also yield something similar to 10.

#### AI Generation Disclosure
Copilot helped with test data generation and using them in the tests.

### CODE REVIEW

File: tests/xsf_tests/test_bessel_functions_right_tail.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    CAPTURE(x, w, ref, rtol, rel_error);
+    REQUIRE(rel_error <= rtol);
+}
+
+TEST_CASE("j1 right tail gh-large-input", "[j1][xsf_tests]") {
+    using test_case = std::tuple<double, double, double>;
+    // Reference values computed with mpmath with 1000 digits of precision:
+    // x = np.logspace(2, 20, num=20)
+    // x, mpmath.besselj(1, x)
```
Comment (fbourgey) [inline tests/xsf_tests/test_bessel_functions_right_tail.cpp]:
```suggestion
    // Reference values computed with mpmath with 1000 digits of precision:
    // import mpmath as mp
    // import numpy as np

    // mp.mp.dps = 1000

    // xs = np.logspace(2, 20, num=20)
    // ys = [mp.besselj(1, mp.mpf(str(x))) for x in xs]

   // for x, y in zip(xs, ys):
   //     print(f"x = {x}, J1(x) = {mp.nstr(y, 30)}")
```

I suggest providing the full Python script, as the current code would fail because `x` is a NumPy array. Feel free to modify.

File: tests/xsf_tests/test_bessel_functions_right_tail.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    auto [x, ref, rtol] = GENERATE(
+        test_case{100.0, -0.07714535201411216, 1e-15}, test_case{885.8667904100823, -0.020092316011146107, 1e-15},
```
Comment (fbourgey) [inline tests/xsf_tests/test_bessel_functions_right_tail.cpp]: Can you add a few tests around the cutoff value of 10 just in case? Maybe use the same boundary cases used for j0/y0: `nextafter(10.0, 0.0), 10.0, and nextafter(10.0,inf).`

**Comment by fbourgey** [comment]
FYI, Codex confirmed that 10 remains a reasonable cutoff

<details>

```python
import math
  import mpmath as mp

  mp.mp.dps = 100

  # Cephes j1 tail coefficients from include/xsf/cephes/j1.h
  PP = [
      7.62125616208173112003e-4,
      7.31397056940917570436e-2,
      1.12719608129684925192e0,
      5.11207951146807644818e0,
      8.42404590141772420927e0,
      5.21451598682361504063e0,
      1.00000000000000000254e0,
  ]
  PQ = [
      5.71323128072548699714e-4,
      6.88455908754495404082e-2,
      1.10514232634061696926e0,
      5.07386386128601488557e0,
      8.39985554327604159757e0,
      5.20982848682361821619e0,
      9.99999999999999997461e-1,
  ]
  QP = [
      5.10862594750176621635e-2,
      4.98213872951233449420e0,
      7.58238284132545283818e1,
      3.66779609360150777800e2,
      7.10856304998926107277e2,
      5.97489612400613639965e2,
      2.11688757100572135698e2,
      2.52070205858023719784e1,
  ]
  QQ = [
      7.42373277035675149943e1,
      1.05644886038262816351e3,
      4.98641058337653607651e3,
      9.56231892404756170795e3,
      7.99704160447350683650e3,
      2.82619278517639096600e3,
      3.36093607810698293419e2,
  ]


  def polevl(x, c):
      r = 0.0
      for a in c:
          r = r * x + a
      return r


  def p1evl(x, c):
      r = x + c[0]
      for a in c[1:]:
          r = r * x + a
      return r


  def vals(x):
      w = 5.0 / x
      z = w * w
      p = polevl(z, PP) / polevl(z, PQ)
      q = polevl(z, QP) / p1evl(z, QQ)

      old = (
          (p * math.cos(x - 3 * math.pi / 4) - w * q * math.sin(x - 3 * math.pi / 4))
          * math.sqrt(2 / math.pi)
          / math.sqrt(x)
      )

      new = (
          ((w * q - p) * math.cos(x) + (p + w * q) * math.sin(x))
          * math.sqrt(1 / math.pi)
          / math.sqrt(x)
      )

      return old, new


  def rel(y, r):
      return float(abs(mp.mpf(y) - r) / abs(r)) if r else float(abs(y))


  for lo, hi, n in [
      (5, 10, 1000),
      (10, 100, 1000),
      (1e4, 1e5, 1000),
      (1e5, 1e6, 500),
      oldwins = newwins = 0
      worst_old = worst_new = None

      for i in range(n):
          x = lo + (hi - lo) * (i + 0.5) / n
          r = mp.besselj(1, x)
          old, new = vals(x)
          eo = rel(old, r)
          en = rel(new, r)


      print(
          f"{lo:g}-{hi:g}: "
          f"oldmax {oldmax:.2e}@{worst_old:.6g} "
          f"newmax {newmax:.2e}@{worst_new:.6g} "
```

gives

```python
5-10: oldmax 4.72e-14@7.0175 newmax 2.95e-14@7.0175 oldwins 226 newwins 386
10-100: oldmax 8.53e-11@69.895 newmax 1.89e-13@69.895 oldwins 31 newwins 933
100-1000: oldmax 6.21e-11@761.05 newmax 6.78e-14@761.05 oldwins 8 newwins 992
1000-10000: oldmax 1.40e-10@8159.5 newmax 7.51e-14@4964.5 oldwins 0 newwins 1000
10000-100000: oldmax 2.42e-10@36415 newmax 1.52e-14@23635 oldwins 0 newwins 1000
100000-1e+06: oldmax 1.66e-08@408700 newmax 2.31e-14@408700 oldwins 0 newwins 500
1e+06-1e+08: oldmax 9.62e-07@8.6965e+07 newmax 3.51e-15@8.6965e+07 oldwins 0 newwins 300
```

</details>

**Comment by dschmitz89** [comment]
> FYI, Codex confirmed that 10 remains a reasonable cutoff
> 
> Details

Thanks!

File: tests/xsf_tests/test_bessel_functions_right_tail.cpp (hunk 2 above)
Comment (dschmitz89) [reply tests/xsf_tests/test_bessel_functions_right_tail.cpp]: One thing we have not tested so far akso for j0 and y0 are huge negative values. Is that unnecessary?

**Comment by dschmitz89** [comment]
@fbourgey Could you give a final approval if this looks good to you now? I will squash merge, the individual commits are not helpful here.

File: tests/xsf_tests/test_bessel_functions_right_tail.cpp [hunk 3]
```diff
@@ -58,19 +58,26 @@ TEST_CASE("y0 right tail gh-large-input", "[y0][xsf_tests]") {
...
-    //     print(f"x = {x}, J1(x) = {mp.nstr(y, 30)}")
+    for x in x_cases:
+        y = mp.besselj(1, mp.mpf(str(x)))
+        print(f"x = {x}, J1(x) = {float(y)}")
+    */
     auto [x, ref, rtol] = GENERATE(
-        test_case{100.0, -0.07714535201411216, 1e-15}, test_case{885.8667904100823, -0.020092316011146107, 1e-15},
+        test_case(std::nextafter(10, 100), 0.043472746168860994, 5e-15),
+        test_case(std::nextafter(10, 0.0), 0.04347274616886188, 5e-15), test_case{100.0, -0.07714535201411216, 1e-15},
+        test_case{885.8667904100823, -0.020092316011146107, 1e-15},
```
Comment (fbourgey) [inline tests/xsf_tests/test_bessel_functions_right_tail.cpp:80]: nit: if I am being picky, we could also test the cutoff 10 to match the y0/j0 case.

**Review by fbourgey — COMMENTED** [review]
LGTM @dschmitz89!

File: tests/xsf_tests/test_bessel_functions_right_tail.cpp (hunk 3 above)
Comment (dschmitz89) [reply tests/xsf_tests/test_bessel_functions_right_tail.cpp:80]: For the next PR :)
